Add support for a `Range` argument to `String#byteslice`

lopopolo opened the pull request:
Add support for `Range` arguments in `String#byteslice` of the form:

```ruby
s = "abc"
raise unless s.byteslice(1..2) == "bc"
```

As of this commit all `String#byteslice` specs pass except for two which
require the `begin` and `end` components of the `Range` to go through
the "implicitly convert to int" flow. Range parsing is still implemented
in mruby which does not do implicit conversions.

Getting the specs to pass required significant rework of the
`protect::is_range` function to distinguish between the "range out" case
and the "valid" case. For example:

```ruby
"".byteslice(-1..-1)
```

was previously squashing this range into the `None` case, which led to
`byteslice` blowing up further down the routine when attempting to fall
back to the integer argument path.

This change led to re-evaluating all other instances of range argument
handling and corrected the response when the "out of range" case is
encountered. The most significant change was in `Array#[]=` which raises
an exception in this case.

Fixes #1916.

Among the core specs:

Before:
Passed 3944, skipped 1063, not implemented 724, failed 4494 specs.

After:
~~Passed 4014, skipped 1065, not implemented 724, failed 4561 specs.~~
Passed 4051, skipped 1075, not implemented 746, failed 4469 specs.

The relevant mruby code is here:

https://github.com/artichoke/artichoke/blob/f86225f2256b51939d7a2ea0c73ddb36a99f64ab/artichoke-backend/vendor/mruby/src/range.c#L456-L488

lopopolo committed:
Add support for a `Range` argument to `String#byteslice`

Add support for `Range` arguments in `String#byteslice` of the form:

```ruby
s = "abc"
raise unless s.byteslice(1..2) == "bc"
```

As of this commit all `String#byteslice` specs pass except for two which
require the `begin` and `end` components of the `Range` to go through
the "implicitly convert to int" flow. Range parsing is still implemented
in mruby which does not do implicit conversions.

Getting the specs to pass required significant rework of the
`protect::is_range` function to distinguish between the "range out" case
and the "valid" case. For example:

```ruby
"".byteslice(-1..-1)
```

was previously squashing this range into the `None` case, which led to
`byteslice` blowing up further down the routine when attempting to fall
back to the integer argument path.

This change led to re-evaluating all other instances of range argument
handling and corrected the response when the "out of range" case is
encountered. The most significant change was in `Array#[]=` which raises
an exception in this case.

Fixes #1916.

Among the core specs:

Before:
Passed 3944, skipped 1063, not implemented 724, failed 4494 specs.

After:
Passed 4014, skipped 1065, not implemented 724, failed 4561 specs.

lopopolo commented on artichoke-backend/src/extn/core/string/mruby.rs:
```diff
@@ -213,7 +213,17 @@ unsafe extern "C" fn string_byteslice(mrb: *mut sys::mrb_state, slf: sys::mrb_va
+            let target_rbasic = value.value.p.cast::<sys::RBasic>();
+
+            // Copy `RClass` from source class to newly allocated `Array`.
+            (*target_rbasic).c = rclass;
```
to get the specs to pass, `String#byteslice` needs to return slices as a the same class as the receiver.

All `String` and `String` subclasses are backed by `spinoso_string::String` and have type tag of `MRB_TT_STRING`, but they may vary in their `*mut RClass` pointer.

`String::alloc_value` by default hardcodes the `String` class pointer, so we fix it up here.

lopopolo commented on artichoke-backend/src/extn/core/string/mruby.rs:
```diff
@@ -213,7 +213,17 @@ unsafe extern "C" fn string_byteslice(mrb: *mut sys::mrb_state, slf: sys::mrb_va
     let result = trampoline::byteslice(&mut guard, value, index, length);
     match result {
-        Ok(value) => value.inner(),
+        Ok(value) if value.is_nil() => value.inner(),
```
`nil` values are structured differently and if we don't short circuit here, attempting to set the class pointer on a `nil` will segfault.

lopopolo reviewed (commented):
@b-n it'd probably be easiest to review this by suppressing whitespace changes in the diff. A lot of the LOC changed are reindenting due to the new match structure as opposed to `if let ... else`.

lopopolo commented on artichoke-backend/src/extn/core/array/args.rs:
```diff
@@ -20,23 +20,31 @@ pub fn element_reference(
+    let rangelen = i64::try_from(ary_len).map_err(|_| Fatal::from("Range length exceeds Integer max"))?;
+    match elem.is_range(interp, rangelen)? {
+        None => {
+            let index = implicitly_convert_to_int(interp, elem)?;
```
moving the implicit conversion into the "this is not a range" case allows us to properly propagate the underlying `TypeError`s.

lopopolo committed:
Fixup array args handling

Move implicit conversion to int into "not a range" branch and propagate
implicit conversion errors.

lopopolo committed:
Fix ruby formatting

b-n commented on artichoke-backend/src/extn/core/string/string_test.rb:
```diff
@@ -47,6 +48,18 @@ def string_byteslice
   raise unless s.byteslice(5, -10).nil?
   raise unless s.byteslice(5, -2).nil?
+  # range
+  raise unless s.byteslice(1..4) == 'bcde'
+  raise unless s.byteslice(1..-1) == 'bcdefghijk'
+  raise unless s.byteslice(10..-1) == 'k'
+  raise unless s.byteslice(20..-1).nil?
+  raise unless s.byteslice(20..-20).nil?
+  raise unless s.byteslice(2..-20) == ''
+  raise unless s.byteslice(-1..20) == 'k'
+  raise unless s.byteslice(-20..20).nil?
+  raise unless s.byteslice(-5..-1) == 'ghijk'
+  raise unless s.byteslice(-5..1) == ''
+  raise unless s.byteslice(-5..8) == 'ghi'
```
I think I've figured out all the permutations of ranges that could be provided (and their `irb` values):

`s` =  start
`e` =  end
`S` = -start
`E` = -end
`x` =  start/end same value
`X` = -start/-end same value
`*` =  same element, different signs, see range

```plain
  "abcdefghijk".byteslice(range)
   |---------|   

# Normal ranges
   x---------|   =   0..0   | "a"
   se--------|   =   0..1   | "ab"
   s---------e   =   0..10  | "abcdefghijk"
   |s-------e|   =   1..9   | "bcdefghij"
   |--------se   =   9..10  | "jk"
   |--------s-e  =   9..11  | "jk"
   |---------x   =  10..10  | "k"
   |---------se  =  10..11  | "k"
   |---------|x  =  11..11  | ""
   |---------|se =  11..12  | ""
  
# Inverse ranges
   es--------|   =   1..0   | "" 
   e---------s   =  10..0   | "" 
   |e-------s|   =   9..1   | ""
   |--------es   =  10..9   | ""
   |--------e-s  =  11..9   | ""
   |---------es  =  11..10  | ""
   |---------|es =  12..11  | nil
  
# S-Negative normal ranges
  Se---------|   = -12..0   | nil 
  S-e--------|   = -12..1   | nil
   *---------|   = -11..0   | "a"
   Se--------|   = -11..1   | "ab"
   S---------e   = -11..10  | "abcdefghijk"
   S----------e  = -11..11  | "abcdefghijk"
   |S-------e|   = -10..9   | "bcdefghij"
   |--------Se   =  -2..10  | "jk"
   |---------*   =  -1..10  | "k"
  
# E-Negative normal ranges
   *---------|   =   0..-11 | "a"
   sE--------|   =   0..-10 | "ab"
   s---------E   =   0..-1  | "abcdefghijk"
   |s-------E|   =   1..-2  | "bcdefghij"
   |--------sE   =   9..-1  | "jk"
   |---------*   =  10..-1  | "k"

# S-Negative and E-Negative inverse ranges
  Es---------|   =   0..-12 | ""
  E-s--------|   =   1..-12 | ""
   Es--------|   =   1..-11 | ""
   |--------Es   =  10..-2  | ""
   |--------E-s  =  11..-2  | ""
   |---------Es  =  11..-1  | ""

# Dbl-Negative ranges
 SE|---------|   = -13..-12 | nil
  X|---------|   = -12..-12 | nil
  SE---------|   = -12..-11 | nil
  S-E--------|   = -12..-10 | nil
   X---------|   = -11..-11 | "a"
   SE--------|   = -11..-10 | "ab"
   S---------E   = -11..-1  | "abcdefghijk"
   |--------SE   =  -2..-1  | "jk"
   |---------X   =  -1..-1  | "k"

# Dbl-Negative inverse ranges
 ES|---------|   = -12..-13 | nil
  ES---------|   = -11..-12 | ""
  E-S--------|   = -10..-12 | ""
   ES--------|   = -10..-11 | ""
   E---------S   =  -1..-11 | ""
   |--------ES   =  -1..-2  | ""
```

I think there might be some cases that can be added if you're really keen :smile:

b-n replied:
This is the same as the above table, just condensed into the ruby tests (passes on MRI 3.1.2)
```suggestion
  raise unless s.byteslice(0..0) == "a"
  raise unless s.byteslice(0..1) == "ab"
  raise unless s.byteslice(0..10) == "abcdefghijk"
  raise unless s.byteslice(1..9) == "bcdefghij"
  raise unless s.byteslice(9..10) == "jk"
  raise unless s.byteslice(9..11) == "jk"
  raise unless s.byteslice(10..10) == "k"
  raise unless s.byteslice(10..11) == "k"
  raise unless s.byteslice(11..11) == ""
  raise unless s.byteslice(11..12) == ""
  raise unless s.byteslice(1..0) == "" 
  raise unless s.byteslice(10..0) == "" 
  raise unless s.byteslice(9..1) == ""
  raise unless s.byteslice(10..9) == ""
  raise unless s.byteslice(11..9) == ""
  raise unless s.byteslice(11..10) == ""
  raise unless s.byteslice(12..11).nil?
  raise unless s.byteslice(-12..0).nil? 
  raise unless s.byteslice(-12..1).nil?
  raise unless s.byteslice(-11..0) == "a"
  raise unless s.byteslice(-11..1) == "ab"
  raise unless s.byteslice(-11..10) == "abcdefghijk"
  raise unless s.byteslice(-11..11) == "abcdefghijk"
  raise unless s.byteslice(-10..9) == "bcdefghij"
  raise unless s.byteslice(-2..10) == "jk"
  raise unless s.byteslice(-1..10) == "k"
  raise unless s.byteslice(0..-11) == "a"
  raise unless s.byteslice(0..-10) == "ab"
  raise unless s.byteslice(0..-1) == "abcdefghijk"
  raise unless s.byteslice(1..-2) == "bcdefghij"
  raise unless s.byteslice(9..-1) == "jk"
  raise unless s.byteslice(10..-1) == "k"
  raise unless s.byteslice(0..-12) == ""
  raise unless s.byteslice(1..-12) == ""
  raise unless s.byteslice(1..-11) == ""
  raise unless s.byteslice(10..-2) == ""
  raise unless s.byteslice(11..-2) == ""
  raise unless s.byteslice(11..-1) == ""
  raise unless s.byteslice(-13..-12).nil?
  raise unless s.byteslice(-12..-12).nil?
  raise unless s.byteslice(-12..-11).nil?
  raise unless s.byteslice(-12..-10).nil?
  raise unless s.byteslice(-11..-11) == "a"
  raise unless s.byteslice(-11..-10) == "ab"
  raise unless s.byteslice(-11..-1) == "abcdefghijk"
  raise unless s.byteslice(-2..-1) == "jk"
  raise unless s.byteslice(-1..-1) == "k"
  raise unless s.byteslice(-12..-13).nil?
  raise unless s.byteslice(-11..-12) == ""
  raise unless s.byteslice(-10..-12) == ""
  raise unless s.byteslice(-10..-11) == ""
  raise unless s.byteslice(-1..-11) == ""
  raise unless s.byteslice(-1..-2) == ""
```

b-n commented on artichoke-backend/src/sys/protect.rs:
```diff
@@ -148,10 +148,10 @@ pub unsafe fn is_range(
-    pub len: sys::mrb_int,
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub enum Range {
+    Valid { start: sys::mrb_int, len: sys::mrb_int },
```
I'm assuming that the start value is enforced to be an `sys::mrb_int`?

```
irb(main):001:0> 1.5..3
=> 1.5..3
```

^-- this is MRI 2.6.3, but gives  the same on MRI 3.1.2. I never really thought of ranges having fractional parts, but apparently that's the MRI world.

lopopolo replied:
When `Array`s and `String`s accept range arguments, they require them to be ints.

The mruby `mrb_range_beg_len` requires `start` and `end` to be ints:

https://github.com/artichoke/artichoke/blob/7bf3f033e020dd60f4b5fd5974dcb19af2617f6d/artichoke-backend/vendor/mruby/src/range.c#L466-L467

b-n commented on artichoke-backend/src/sys/protect.rs:
```diff
@@ -164,20 +164,28 @@ struct IsRange {
-            sys::mrb_sys_cptr_value(mrb, out.cast::<c_void>())
-        } else {
-            sys::mrb_sys_nil_value()
+        match check_range {
```
Am I interpreting this correctly, does mruby change a Range into a start + len for us? That's pretty nice.

e.g.
```
a = "hello"
a.byteslice(-3..-3)
```

In the above, `sys::mrb_range_beg_len` would return `start: 2, len: 1`? Very nice if it does

lopopolo replied:
yes that's correct. The routine looks like this:

https://github.com/artichoke/artichoke/blob/7bf3f033e020dd60f4b5fd5974dcb19af2617f6d/artichoke-backend/vendor/mruby/src/range.c#L456-L488

b-n commented on artichoke-backend/src/extn/core/string/trampoline.rs:
```diff
@@ -410,6 +418,74 @@ pub fn byteslice(
+            };
+            let index = match index {
+                None => return Ok(Value::nil()),
+                Some(index) if index > s.len() => return Ok(Value::nil()),
```
I think my understanding of how mruby unwraps into `Range::Valid` isn't clear. Should mruby take care of these checks and provide a `Range::Out` instead?

lopopolo replied:
I think you're probably right let me check this case in the repl ...

```
artichoke 0.1.0-pre.0 (2022-07-23 revision 5844) [x86_64-apple-darwin]
[rustc 1.62.1 (e092d0b6b 2022-07-16) on x86_64-apple-darwin]
>>> s = "Abc"
=> "Abc"
>>> s.byteslice(4..6)
[artichoke-backend/src/extn/core/string/trampoline.rs:440] Value::nil() = Value(
    nil,
)
=> nil
>>> s.byteslice(10..20)
[artichoke-backend/src/extn/core/string/trampoline.rs:440] Value::nil() = Value(
    nil,
)
=> nil
```

Somehow this case is getting hit.

lopopolo replied:
After taking a peek at the code, the given length argument is only used for wrapping negative start and end of the `Range`.

As long as these come out postive, regardless of whether they overlap with `0..len`, `MRB_RANGE_OK` is returned.

https://github.com/artichoke/artichoke/blob/7bf3f033e020dd60f4b5fd5974dcb19af2617f6d/artichoke-backend/vendor/mruby/src/range.c#L456-L488

b-n reviewed (approved):
Pretty nice refactorings. I've got a couple of questions around the `sys::mrb_range_beg_len` and a couple of tests for byteslice that could be added.

I'm wondering whether something like this would ever make sense:

```rust
impl for Range {
    pub fn array_accessor(interp: &mut Artichoke, cap: Option<Value>) -> (usize, usize) {
        match Range {
            Range::Valid { start, len } => {
                  //do all the checking of start/len against the cap
                  (start, len)
            },
            _ => Err()
        }
    }
}
```

:point_up: Something that does all the conversion to usize, and just returns the elements to pluck out of a list. I don't think there's anything particular special about a range, except it gives you many ways to access elements in an array in this case.

I'm approving since I don't see anything in particular that could/should break.

lopopolo committed:
Update artichoke-backend/src/extn/core/string/string_test.rb

Co-authored-by: Ben Naylor <nayben@gmail.com>

lopopolo commented on artichoke-backend/src/extn/core/string/trampoline.rs:
```diff
@@ -410,6 +418,74 @@ pub fn byteslice(
 ) -> Result<Value, Error> {
     let s = unsafe { super::String::unbox_from_value(&mut value, interp)? };
+    let maybe_range = if length.is_none() {
+        index.is_range(interp, s.char_len() as i64)?
```
```suggestion
        index.is_range(interp, s.len() as i64)?
```

since we're doing byte slices.

lopopolo commented:
> 

Great idea! Filed https://github.com/artichoke/artichoke/issues/1971.

lopopolo committed:
Format Ruby sources with RuboCop

lopopolo committed:
Add `String#byteslice` tests for string where length != bytesize

These tests fail:

```
running 1 test
test extn::core::string::tests::functional ... FAILED

failures:

---- extn::core::string::tests::functional stdout ----
thread 'extn::core::string::tests::functional' panicked at '
String tests failed with RuntimeError exception
    message: ""
    backtrace:
        (eval):126:in string_byteslice
        (eval):8:in spec
        (eval):1
', artichoke-backend/src/extn/core/string/mod.rs:164:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

lopopolo committed:
Fix `String#bytesize` with `Range` argument for multi-byte strings

lopopolo commented:
Added some tests with multibyte UTF-8 character sequence to cover the `char_len` -> `bytesize` fix.
